launcher/modManager/chroniclesextractor.cpp

```diff
@@ -0,0 +1,276 @@
+		return 0;
+	}
+
+	QByteArray dataBegin = file.read(1'000'000);
+	int chronicle = 0;
+	for (const auto& kv : chronicles) {
+		if(dataBegin.contains(kv.second))
+		{
+			chronicle = kv.first;
+			break;
+		}
+	}
+	if(!chronicle)
+	{
+		QMessageBox::critical(parent, tr("Invalid file selected"), tr("You have to select an chronicle installer file!"));
+		return 0;
+	}
+	return chronicle;
+}
+
+bool ChroniclesExtractor::extractGogInstaller(QString file)
+{
+#ifndef ENABLE_INNOEXTRACT
+		QMessageBox::critical(parent, tr("Innoextract functionality missing"), "VCMI was compiled without innoextract support, which is needed to extract chroncles!");
+		return false;
+#else
+		::extract_options o;
+		o.extract = true;
+
+		// standard settings
+		o.gog_galaxy = true;
+		o.codepage = 0U;
+		o.output_dir = tempDir.path().toStdString();
+		o.extract_temp = true;
+		o.extract_unknown = true;
+		o.filenames.set_expand(true);
+
+		o.preserve_file_times = true; // also correctly closes file -> without it: on Windows the files are not written completely
+
+		QString errorText = "";
+		try
+		{
+			process_file(file.toStdString(), o, [this](float progress) {
+				float overallProgress = ((1.0 / static_cast<float>(fileCount)) * static_cast<float>(extractionFile)) + (progress / static_cast<float>(fileCount));
+				if(cb)
+					cb(overallProgress);
+			});
+		}
+		catch(const std::ios_base::failure & e)
+		{
+			errorText = tr("Stream error while extracting files!\nerror reason: ");
+			errorText += e.what();
+		}
+		catch(const format_error & e)
+		{
+			errorText = e.what();
+		}
+		catch(const std::runtime_error & e)
+		{
+			errorText = e.what();
+		}
+		catch(const setup::version_error &)
+		{
+			errorText = tr("Not a supported Inno Setup installer!");
+		}
+
+		if(!errorText.isEmpty())
+		{
+			QMessageBox::critical(parent, tr("Extracting error!"), errorText);
+			return false;
+		}
+
+		return true;
+#endif
+}
+
+void ChroniclesExtractor::createBaseMod() const
+{
+	QDir dir(pathToQString(VCMIDirs::get().userDataPath() / "Mods"));
+	dir.mkdir("chronicles");
+	dir.cd("chronicles");
+	dir.mkdir("Mods");
+
+	QJsonObject mod
+	{
+		{ "modType", "Expansion" },
+		{ "name", "Heroes Chronicles" },
+		{ "description", "Heroes Chronicles" },
+		{ "author", "3DO" },
+		{ "version", "1.0" },
+		{ "contact", "vcmi.eu" },
+	};
+
+	QFile jsonFile(dir.filePath("mod.json"));
+    jsonFile.open(QFile::WriteOnly);
+    jsonFile.write(QJsonDocument(mod).toJson());
+}
+
+void ChroniclesExtractor::createChronicleMod(int no)
+{
+	QDir dir(pathToQString(VCMIDirs::get().userDataPath() / "Mods" / "chronicles" / "Mods" / ("chronicles_" + std::to_string(no))));
+	dir.removeRecursively();
+	dir.mkpath(".");
+
+	QJsonObject mod
+	{
+		{ "modType", "Expansion" },
+		{ "name", "Heroes Chronicles - " + QString::number(no) },
+		{ "description", "Heroes Chronicles - " + QString::number(no) },
```

Comment:
Perhaps use chronicle name, e.g. "World Tree" instead? Might be too long for mod name, but not a problem for description

Reply:
Oh, or remove "Heroes Chronicles" part and name them directly, e.g. "World Tree" - this is submod so it will be visible only inside of "Heroes Chronicles" submod.
No need for 8 submods with almost identical name